idx/memory/tag_query.go

```diff
@@ -171,247 +240,728 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
-					if _, ok := matchingTags[value]; ok {
-						if not {
-							delete(resultSet, id)
-						}
-						continue IDS
-					}
+			// reduce regex matching by looking up cached matches
+			if _, ok := e.matchCache.Load(value); ok {
+				if not {
+					return false
+				}
+				continue EXPRS
+			}
 
-					// value == nil means that this expression can be short cut
-					// by not evaluating it
-					if e.value == nil || e.value.MatchString(value) {
-						if len(matchingTags) < matchCacheSize {
-							matchingTags[value] = struct{}{}
-						}
-						if not {
-							delete(resultSet, id)
-						}
-						continue IDS
-					} else {
-						if len(notMatchingTags) < matchCacheSize {
-							notMatchingTags[value] = struct{}{}
-						}
-					}
+			// value == nil means that this expression can be short cut
+			// by not evaluating it
+			if e.value == nil || e.value.MatchString(value) {
+				if atomic.LoadInt32(&e.matchCacheSize) < int32(matchCacheSize) {
+					e.matchCache.Store(value, struct{}{})
+					atomic.AddInt32(&e.matchCacheSize, 1)
 				}
-				if !not {
-					delete(resultSet, id)
+				if not {
+					return false
+				}
+				continue EXPRS
+			} else {
+				if atomic.LoadInt32(&e.missCacheSize) < int32(matchCacheSize) {
+					e.missCache.Store(value, struct{}{})
+					atomic.AddInt32(&e.missCacheSize, 1)
 				}
 			}
 		}
+		if !not {
+			return false
+		}
 	}
+	return true
 }
 
-func (q *TagQuery) filterByFrom(resultSet TagIDs, byId map[string]*idx.Archive) {
-	if q.from <= 0 {
-		return
+// testByTagMatch filters a given metric by matching a regular expression against
+// the associated tags
+func (q *TagQuery) testByTagMatch(def *idx.Archive) bool {
+	for _, tag := range def.Tags {
+		equal := strings.Index(tag, "=")
+		if equal < 0 {
+			corruptIndex.Inc()
+			log.Error(3, "memory-idx: tag is in index, but does not contain '=' sign: %s", tag)
+			continue
+		}
+		key := tag[:equal]
+
+		if _, ok := q.tagMatch.missCache.Load(key); ok {
+			continue
+		}
+
+		if _, ok := q.tagMatch.matchCache.Load(key); ok || q.tagMatch.value.MatchString(key) {
+			if !ok {
+				if atomic.LoadInt32(&q.tagMatch.matchCacheSize) < int32(matchCacheSize) {
+					q.tagMatch.matchCache.Store(key, struct{}{})
+					atomic.AddInt32(&q.tagMatch.matchCacheSize, 1)
+				}
+			}
+			return true
+		} else {
+			if _, ok := q.tagMatch.missCache.Load(key); !ok {
+				if atomic.LoadInt32(&q.tagMatch.missCacheSize) < int32(matchCacheSize) {
+					q.tagMatch.missCache.Store(key, struct{}{})
+					atomic.AddInt32(&q.tagMatch.missCacheSize, 1)
+				}
+			}
+			continue
+		}
+	}
+
+	return false
+}
+
+// testByFrom filters a given metric by its LastUpdate time
+func (q *TagQuery) testByFrom(def *idx.Archive) bool {
+	return q.from <= def.LastUpdate
+}
+
+// testByPrefix filters a given metric by matching prefixes against the values
+// of a specific tag
+func (q *TagQuery) testByPrefix(def *idx.Archive, exprs []kv) bool {
```

Comment:
This function doesn't handle the `name` special case like testbyMatch does.

Reply:
this is better: https://github.com/grafana/metrictank/pull/779/commits/a1902cfb78165f6572b5be2c4ca90ee4aa6deab1

Reply:
right, thanks!

Reply:
that should be better: https://github.com/grafana/metrictank/pull/779/commits/544a43572e439b68eb0e04877d0e75844140b412